server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +27,118 @@ import org.http4s.Request
+        for {
+          _ <- testee.takeToken
+          unavailable <- testee.takeToken
+          _ <- IO.sleep(someRefillTime)
+          available <- testee.takeToken
+        } yield unavailable -> available
+      }
 
-    val takeTokenAfterRefill = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(101.milliseconds) *>
-        testee.takeToken
+      TestControl
+        .executeEmbed(takeTokenAfterRefill)
+        .map { case (a, b) => a.isInstanceOf[TokenUnavailable] && b == TokenAvailable }
+        .assert
     }
-
-    takeTokenAfterRefill
-      .map { result =>
-        ctx.advanceAndTick(101.milliseconds)
-        result
-      }
-      .assertEquals(TokenAvailable)
   }
 
   test("LocalTokenBucket should not add another token at specified interval when at capacity") {
-    val ctx = TestContext()
-    val capacity = 5
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
-
-    val takeExtraToken = createBucket.flatMap { testee =>
-      val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
-        testee.takeToken
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeExtraToken = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
+          testee.takeToken
+        }
+        takeFiveTokens >> IO.sleep(someRefillTime - 1.milliseconds) >> testee.takeToken
       }
-      IO.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
-    }
 
-    takeExtraToken
-      .map { result =>
-        ctx.advanceAndTick(300.milliseconds)
-        result
-      }
-      .map(_.isInstanceOf[TokenUnavailable])
-      .assert
+      TestControl
+        .executeEmbed(takeExtraToken)
+        .map {
+          case TokenUnavailable(Some(_)) => true
+          case _ => false
+        }
+        .assert
+    }
   }
 
   test(
     "LocalTokenBucket should only return a single token when only one token available and there are multiple concurrent requests"
   ) {
     val capacity = 1
     val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+      TokenBucket.local[IO](capacity, 1.milliseconds)
 
     val takeTokensSimultaneously = createBucket.flatMap { testee =>
       (1 to 5).toList.parTraverse(_ => testee.takeToken)
     }
 
-    takeTokensSimultaneously
+    val prog = takeTokensSimultaneously
       .map { result =>
         result.count(_ == TokenAvailable)
       }
-      .assertEquals(1)
+
+    TestControl.executeEmbed(prog).assertEquals(1)
   }
 
+  val localGen = for {
+    // guarantee a refill time > 0
+    fd1 <- genFiniteDuration
+    // guarantee some wait time < refill time
+    fd2 <- genFiniteDuration.map(i => if (i < fd1) i else fd1 - 1.millisecond)
+  } yield fd1 -> fd2
```

Comment:
Sorry, just a nit, if `fd1` and `fd2` are the refill and wait times, can we just name them that way? :)

Reply:
Yep, of course